### DISCUSSION TITLE
Use kamal for existing images

### DISCUSSION THREAD

**Discussion by AdalyatNazirov** [basecamp/kamal#1132] [Q&A]
What i want to reach is to run existing docker image hosted on digitalocean registry.

I found `--skip-push` command option but still required me to setup builder/arch in config.  Also I'm not sure which configuration parameters I have to use in that case. I think service and image? If that's true, cannot find a port option for this case (i need port to be opened other than 80/443.

Any help or ideas are approciated

**Comment by tuladhar** [comment]
If you're using an existing existing Docker image that wasn't built with Kamal, make sure to add `service` label to your Docker image. This label should match the service in your `config/deploy.yml`.

**Reply by eirvandelden** [reply]
How could one add the service label to an existing Docker image?

I tried adding it to the custom labels:
```yaml
servers:
  web:
    hosts:
      - HOST
    labels:
      service: "MY_APP"
```
But it is still giving the same error:
```
docker stdout: Image REGISTRY/MY_APP:latest is missing the 'service' label
docker stderr: Nothing written
```

**Comment by edolix** [comment]
@AdalyatNazirov did you find a solution for this? Thanks!

**Comment by shuiRong** [comment]
Recently, I solved this issue and successfully managed to skip the Kamal build step by directly using the latest image from Docker Hub. However, I still have a few unverified question, which I’ll list at the end.

Basically, it’s the same approach eirvandelden mentioned earlier. Here are the steps:
1. Modify your project’s **Dockerfile** by adding this line right after the `FROM` instruction:
`LABEL service=<service_name>`
Make sure `<service_name>` matches your actual service name(in `config/deploy.yml`). Then, push this modified image (with the added LABEL) to Docker Hub.
2. Deploy using this command:
`kamal deploy -P --version=latest`
This tells Kamal to skip the local image build and instead pull the image from Docker Hub with the latest tag. Of course, you can modify the (latest)tag to suit your needs.

That's it.

I know this might feel a bit redundant, since you still have to update the Dockerfile, rebuild the image, and push it to Docker Hub. But it works well for my case. That’s because in my case, the build process is handled by a separate CI/CD pipeline that I can’t modify to handle deployments to multiple servers. So I introduced Kamal to handle deployments independently.

> That said, I’m still left with two open questions I haven’t figured out yet, and I’ll just leave them here:
> 1. Is it mandatory for the image deployed by Kamal to have the LABEL service tag? If so, that’s a bit unfortunate, since many images > don’t include this label by default.
> 2. Does the image being deployed have to be under your own Docker Hub account?

@AdalyatNazirov @edolix @eirvandelden I’m not sure if this is still useful to you, but I’ll leave it here in case someone Googles it.

**Reply by nickhammond** [reply]
@shuiRong 

> Is it mandatory for the image deployed by Kamal to have the LABEL service tag? If so, that’s a bit unfortunate, since many images > don’t include this label by default.

Yes, the LABEL is used when calling commands like `docker ps` to find the relevant containers running for that service. I stumbled on that same issue with [the buildpack's PR](https://github.com/basecamp/kamal/pull/916). It makes the most sense to add the label for future use, it's the most applicable way to filter the various docker commands to get the details Kamal needs.

Here's an example of using the existing nginx image and just adding the service label, you would then manage this Dockerfile and this would be what Kamal would deploy. 

```
FROM nginx:alpine

LABEL service="theservice"
```

> Does the image being deployed have to be under your own Docker Hub account?

No, you just need to supply credentials that have access to the image you've specified.

**Reply by edolix** [reply]
Thanks @shuiRong - thanks for sharing the details! I decided to take another approach but glad to read you figured it out! 👍
